lib/calendars.ml

```diff
@@ -251,194 +303,226 @@ let sdn_of_hebrew d =
-           delta = 0}
-    else sdn + 1
+  let hebrew_nb_days_upper_bound =
+    (* last two are for Adar I and II;
+       Adar II should be 29, but we keep it simple and do not check for leap years *)
+    [| 30; 29; 30; 29; 30; 29; 30; 30; 30; 29; 30; 30; 30 |]
   in
-  let d1 = f sdn in
-  let d2 = f (sdn_max + d.delta) in
-  if d1.day = 1 && d2.day = 1 then
-    if d1.month = 1 && d2.month = 1 then
-      if d1.year + 1 = d2.year then
-        {day = 0; month = 0; year = d1.year; delta = 0}
-      else {d1 with delta = sdn_max + d.delta - sdn - 1}
-    else if
-      d1.month + 1 = d2.month ||
-      d1.month = f_max_month && d1.year + 1 = d2.year
-    then
-      {d1 with day = 0}
-    else {d1 with delta = sdn_max + d.delta - sdn - 1}
-  else {d1 with delta = sdn_max + d.delta - sdn - 1}
-
-let gregorian_of_julian = conv gregorian_of_sdn 12 sdn_of_julian 12
-let julian_of_gregorian = conv julian_of_sdn 12 sdn_of_gregorian 12
-
-let gregorian_of_french = conv gregorian_of_sdn 12 sdn_of_french 13
-let french_of_gregorian = conv french_of_sdn 13 sdn_of_gregorian 12
-
-let gregorian_of_hebrew = conv gregorian_of_sdn 12 sdn_of_hebrew 13
-let hebrew_of_gregorian = conv hebrew_of_sdn 13 sdn_of_gregorian 12
+  let is_invalid =
+    day < 1 || month < 1
+    ||
+    match kind with
+    | Gregorian | Julian ->
+        (* from wikipedia: "A year zero does not exist in the Anno Domini (AD) calendar year
+           system commonly used to number years in the Gregorian calendar (nor in its
+           predecessor, the Julian calendar)" *)
+        year = 0
+        || (match kind with
+           | Gregorian -> true
+           | Julian -> year > -46
+           | Hebrew | French -> (* . *) assert false)
+           && (month > 12 || gregorian_nb_days_upper_bound.(month - 1) < day)
+    | Hebrew -> month > 13 || hebrew_nb_days_upper_bound.(month - 1) < day
+    | French -> month > 12 || day > 30
+  in
+  if is_invalid then
+    Error
+      (Printf.sprintf "Invalid value: day=%d month=%d year=%d delta=%d kind=%s"
+         day month year delta (kind_to_string kind))
+  else Ok { day; month; year; delta; kind }
+
+let to_sdn : type a. a date -> sdn =
+ fun date ->
+  match date.kind with
+  | Gregorian -> sdn_of_gregorian date
+  | Julian -> sdn_of_julian date
+  | French -> sdn_of_french date
+  | Hebrew -> sdn_of_hebrew date
+
+let to_gregorian : type a. a date -> gregorian date =
+ fun date ->
+  match date.kind with
+  | Gregorian -> date
+  | Julian | French | Hebrew -> to_sdn date |> gregorian_of_sdn
+
+let to_julian : type a. a date -> julian date =
+ fun date ->
+  match date.kind with
+  | Julian -> date
+  | Gregorian | French | Hebrew -> to_sdn date |> julian_of_sdn
+
+let to_french : type a. a date -> french date =
+ fun date ->
+  match date.kind with
+  | French -> date
+  | Gregorian | Julian | Hebrew -> to_sdn date |> french_of_sdn
+
+let to_hebrew : type a. a date -> hebrew date =
```

Comment:
Here, as you don't have polymorphic recursion, I believe you can use the lighter syntax:

```ocaml
let to_hebrew (type kind) (date : kind date) : hebrew date =
  match date.kind with ...
```

(same for all the others functions above)